### PR TITLE
Alternate Solution for USACO 1018

### PR DESCRIPTION
**Pull request by DenialRiver1434**
_Place an "x" in the corresponding checkbox if it is done or does not apply to this pull request._

- [x] I have tested my code.
- [x] I have added my solution according to the steps [here](https://usaco.guide/general/adding-solution#steps).
- [x] I have followed the code conventions mentioned [here](https://usaco.guide/general/adding-solution/#code-conventions).
  - I understand that if it is clear that I have not attempted to follow these conventions, my PR will be closed.
  - If changes are requested, I will re-request a review after addressing them.
- [x] I have linked this PR to any issues that it closes.

### CODE REVIEW

File: solutions/gold/usaco-1018.mdx [hunk 1]
```diff
@@ -111,3 +112,92 @@ print(ans, file=open("help.out", "w"))
...
+#include <ext/pb_ds/assoc_container.hpp>
+using namespace __gnu_pbds;
+template <class T>
+using Tree =
+    tree<T, null_type, less<T>, rb_tree_tag, tree_order_statistics_node_update>;
+
+const int MOD = 1e9 + 7;
+
+// Binary exponentation
+int powermod(long long base, long long exp) {
```
Comment (SansPapyrus683) [inline solutions/gold/usaco-1018.mdx:163]: did you use the guide's powermod? i'd like utility functions like this to be consistent

File: solutions/gold/usaco-1018.mdx [hunk 2]
```diff
@@ -111,3 +112,92 @@ print(ans, file=open("help.out", "w"))
...
+	while (exp > 0) {
+		if (exp % 2 == 1)  // if n is odd
+			result = result * base % MOD;
+		base = base * base % MOD;
+		exp /= 2;  // divide by two
+	}
+	return result;
+}
+
+int N, tot = 0;
```
Comment (SansPapyrus683) [inline solutions/gold/usaco-1018.mdx:175]: make n lowercase & tot local

File: solutions/gold/usaco-1018.mdx [hunk 3]
```diff
@@ -111,3 +112,92 @@ print(ans, file=open("help.out", "w"))
...
+vector<pair<int, int>> ranges;
+
+int main() {
+	ifstream fin("help.in");
+	ofstream fout("help.out");
```
Comment (SansPapyrus683) [inline solutions/gold/usaco-1018.mdx:180]:
```suggestion
```

File: solutions/gold/usaco-1018.mdx [hunk 4]
```diff
@@ -111,3 +112,92 @@ print(ans, file=open("help.out", "w"))
...
+		int left = ranges[i].first, right = ranges[i].second;
+
+		// Counting # of previous ranges that don't intersect
+		int unaffiliated = pastend.order_of_key({left, -1});
+
+		tot = (2 * tot) % MOD + powermod(2, unaffiliated);
+		tot %= MOD;
+		pastend.insert({right, i});
+	}
+	fout << tot << endl;
```
Comment (SansPapyrus683) [inline solutions/gold/usaco-1018.mdx:202]:
```suggestion

	ofstream("help.out") << tot << endl;
```

File: solutions/gold/usaco-1018.mdx [hunk 5]
```diff
@@ -111,3 +112,92 @@ print(ans, file=open("help.out", "w"))
...
+	ifstream fin("help.in");
+	ofstream fout("help.out");
+
+	fin >> N;
+	for (int i = 0; i < N; i++) {
+		int a, b;
+		fin >> a >> b;
+		ranges.push_back({a, b});
+	}
+	// Sorting ranges by their left values
```
Comment (SansPapyrus683) [inline solutions/gold/usaco-1018.mdx:188]:
```suggestion

	// Sorting ranges by their left values
```

File: solutions/gold/usaco-1018.mdx [hunk 6]
```diff
@@ -111,3 +112,92 @@ print(ans, file=open("help.out", "w"))
...
+```cpp
+#include <bits/stdc++.h>
+using namespace std;
+
+// Importing indexed sets
+#include <ext/pb_ds/assoc_container.hpp>
+using namespace __gnu_pbds;
+template <class T>
+using Tree =
+    tree<T, null_type, less<T>, rb_tree_tag, tree_order_statistics_node_update>;
```
Comment (SansPapyrus683) [inline solutions/gold/usaco-1018.mdx:158]: hb you make this a codesnip (idk, you can reject this)

**Review by SansPapyrus683 — CHANGES_REQUESTED** [review]
btw can you add some hints? i'd really appreciate if you did ❤️

File: solutions/gold/usaco-1018.mdx (hunk 6 above)
Comment (DenialRiver1434) [reply solutions/gold/usaco-1018.mdx:158]: yeah that's a good idea

File: solutions/gold/usaco-1018.mdx (hunk 1 above)
Comment (DenialRiver1434) [reply solutions/gold/usaco-1018.mdx:163]: it has been changed, I also put it in a code snippet such that everything's uniform

File: solutions/gold/usaco-1018.mdx [hunk 7]
```diff
@@ -111,3 +113,101 @@ print(ans, file=open("help.out", "w"))
...
+
+const int MOD = 1e9 + 7;
+// EndCodeSnip
+
+// BeginCodeSnip{Binary Exponentation}
+long long exp(long long x, long long n) {
```
Comment (SansPapyrus683) [inline solutions/gold/usaco-1018.mdx:164]:
```suggestion
// BeginCodeSnip{Binary Exponentation (from module)}
long long exp(long long x, long long n) {
```

File: solutions/gold/usaco-1018.mdx [hunk 8]
```diff
@@ -111,3 +113,101 @@ print(ans, file=open("help.out", "w"))
...
+		x = x * x % MOD;
+		n /= 2;
+	}
+	return res;
+}
+// EndCodeSnip
+
+int n;
+vector<pair<int, int>> ranges;
+
```
Comment (SansPapyrus683) [inline solutions/gold/usaco-1018.mdx:179]: wait none of these have to be global

**Review by SansPapyrus683 — CHANGES_REQUESTED** [review]

File: solutions/gold/usaco-1018.mdx [hunk 9]
```diff
@@ -5,6 +5,20 @@ title: Help Yourself
...
+<Spoiler title="Hint 2 (solution 1)">
+To avoid overcounting, we only count it if our segment is the beginning of a connected component in any given subset.
+</Spoiler>
+
+<Spoiler title="Hint for solution 2">
+We sort the segments by their left endpoints. We then solve the problem for the first segment, then the first two segments, then the first three segments, and so on.
+
+You will need to use [indexed sets](https://usaco.guide/gold/PURS?lang=cpp#finding-the-k-th-element).
+</Spoiler>
```
Comment (SansPapyrus683) [inline solutions/gold/usaco-1018.mdx:20]: just add hints for a single solution

**Review by SansPapyrus683 — APPROVED** [review]
see envy? this is how you should do things

File: solutions/gold/usaco-1018.mdx [hunk 10]
```diff
@@ -111,3 +121,100 @@ print(ans, file=open("help.out", "w"))
 
 </PySection>
 </LanguageSection>
+</Spoiler>
+
+<Spoiler title="Solution 2">
+
+<Warning>
+
+This solution makes use of [indexed sets](https://usaco.guide/gold/PURS?lang=cpp#finding-the-k-th-element).
```
Comment (ryanchou-dev) [inline solutions/gold/usaco-1018.mdx:130]:
```suggestion
This solution makes use of [indexed sets](/gold/PURS?lang=cpp#finding-the-k-th-element).
```

File: solutions/gold/usaco-1018.mdx [hunk 11]
```diff
@@ -111,3 +121,100 @@ print(ans, file=open("help.out", "w"))
...
+
+</Warning>
+
+## Explanation
+
+We process the segments one-by-one. Let the complexity sum of the the first $k$ segments be $c_k$. We'll use $c_k$ to calcualte $c_{k+1}$ in a process similar to dynamic programming. Our answer would then be $c_N$.
```
Comment (ryanchou-dev) [inline solutions/gold/usaco-1018.mdx:136]:
```suggestion
We process the segments one-by-one. Let the complexity sum of the first $k$ segments be $c_k$. We'll use $c_k$ to calculate $c_{k+1}$ in a process similar to dynamic programming. Our answer would then be $c_N$.
```

File: solutions/gold/usaco-1018.mdx [hunk 12]
```diff
@@ -111,3 +121,100 @@ print(ans, file=open("help.out", "w"))
...
+
+From $c_k$ to $c_{k+1}$, our new set of subsets is,
+
+$$\text{\{the previous set of subsets\} } \cup \text{ \{the previous set of subsets with }segment_{k+1}\text{ added to each of them\}}$$
+
+The sum of the complexities of the previous set of subsets is still $c_k$. However, adding $segment_{k+1}$ to each of them may merge lots of previously disjoint segments (if $segment_{k+1}$ happened to intersect all of them).
+
+To control this, **we first sort the segments by their left endpoints**. This means that each new segment can only either merge itself into a previous segment (without merging previously disjoint segments) or form a new disjoint segment.
+
+We also know that our new segment forms a new disjoint segment if and only if it merges with previous segments that doesn't intersect with it. Therefore, the transition formula is,
```
Comment (ryanchou-dev) [inline solutions/gold/usaco-1018.mdx:146]:
```suggestion
We also know that our new segment forms a new disjoint segment if and only if it merges with previous segments that don't intersect with it. Therefore, the transition formula is,
```

File: solutions/gold/usaco-1018.mdx [hunk 13]
```diff
@@ -111,3 +121,100 @@ print(ans, file=open("help.out", "w"))
...
+
+$c_{k+1} = c_k + 2^{\text{number of previous segments that don't intsersect } segment_k}$
```
Comment (ryanchou-dev) [inline solutions/gold/usaco-1018.mdx:148]: eek it's kinda hard to read a \text this long, maybe you could say something like, if x is the number of prev segments ..., and then just write 2^x  or smtg

File: solutions/gold/usaco-1018.mdx [hunk 14]
```diff
@@ -111,3 +121,100 @@ print(ans, file=open("help.out", "w"))
...
+
+To count the $\text{number of previous segments that don't intsersect } segment_k$, we use an indexed set that stores the right endpoints of previous segments. We can use its order_of_key function to determine the number of previous segments that end before our left endpoint.
```
Comment (ryanchou-dev) [inline solutions/gold/usaco-1018.mdx:150]:
```suggestion
To count the $\text{number of previous segments that don't intersect} segment_k$, we use an indexed set that stores the right endpoints of previous segments. We can use its `order_of_key` function to determine the number of previous segments that end before our left endpoint.
```

File: solutions/gold/usaco-1018.mdx [hunk 15]
```diff
@@ -111,3 +121,100 @@ print(ans, file=open("help.out", "w"))
...
+
+## Implementation
+
```
Comment (ryanchou-dev) [inline solutions/gold/usaco-1018.mdx:153]: time complexity?

File: solutions/gold/usaco-1018.mdx [hunk 16]
```diff
@@ -111,3 +121,100 @@ print(ans, file=open("help.out", "w"))
...
+using namespace std;
+
+// BeginCodeSnip{Importing indexed sets}
+#include <ext/pb_ds/assoc_container.hpp>
+using namespace __gnu_pbds;
+template <class T>
+using Tree =
+    tree<T, null_type, less<T>, rb_tree_tag, tree_order_statistics_node_update>;
+
+const int MOD = 1e9 + 7;
```
Comment (ryanchou-dev) [inline solutions/gold/usaco-1018.mdx:168]: move this out?

**Review by ryanchou-dev — CHANGES_REQUESTED** [review]
maybe rename segment to seg and texttt it?

very good

File: solutions/gold/usaco-1018.mdx (hunk 11 above)
Comment (jessechoe10) [inline solutions/gold/usaco-1018.mdx:136]:
```suggestion
Let the complexity sum of the first $k$ segments be $c_k$. Use $c_k$ to calculate $c_{k+1}$ in a process similar to dynamic programming, and our final answer would be $c_N$.
```

File: solutions/gold/usaco-1018.mdx (hunk 12 above)
Comment (jessechoe10) [inline solutions/gold/usaco-1018.mdx:146]:
```suggestion
The sum of the complexities of the previous set of subsets is still $c_k$. However, adding $segment_{k+1}$ to each of them may merge lots of previously disjoint segments (for example, if $segment_{k+1}$ happened to intersect all of them), so **sort the segments by their left endpoints**. 

Each new segment can merge into a previous segment (without merging previously disjoint segments) or form a new disjoint segment, and our new segment forms a new disjoint segment if and only if it merges with previous segments that don't intersect. Therefore, the transition formula is,
```

File: solutions/gold/usaco-1018.mdx (hunk 14 above)
Comment (jessechoe10) [inline solutions/gold/usaco-1018.mdx:150]:
```suggestion
To count the $\text{number of previous segments that don't intersect} segment_k$, use an indexed set that stores the right endpoints of previous segments. Use the `order_of_key` function to determine the number of previous segments that end before our left endpoint.
```

**Review by jessechoe10 — CHANGES_REQUESTED** [review]

File: solutions/gold/usaco-1018.mdx [hunk 17]
```diff
@@ -111,3 +117,103 @@ print(ans, file=open("help.out", "w"))
...
+
+<Warning>
+
+This solution makes use of [indexed sets](/gold/PURS?lang=cpp#finding-the-k-th-element).
+
+</Warning>
+
+## Explanation
+
+Let the complexity sum of the first $k$ segments be $c_k$. Use $c_k$ to calculate $c_{k+1}$ in a process similar to dynamic programming, and our final answer would be $c_N$.
```
Comment (jessechoe10) [inline solutions/gold/usaco-1018.mdx:132]:
```suggestion
Let the complexity sum of the first $k$ segments be $c_k$. Use $c_k$ to calculate $c_{k+1}$ in a process similar to dynamic programming, and our final answer is $c_N$.
```

File: solutions/gold/usaco-1018.mdx [hunk 18]
```diff
@@ -111,3 +117,103 @@ print(ans, file=open("help.out", "w"))
...
+
+From $c_k$ to $c_{k+1}$, our new set of subsets is,
+
+$$\text{\{the previous set of subsets\} } \cup \text{ \{the previous set of subsets with }\texttt{seg}_{k+1}\text{ added to each of them\}}$$
+
+The sum of the complexities of the previous set of subsets is still $c_k$. However, adding $\texttt{seg}_{k+1}$ to each of them may merge lots of previously disjoint segments (if $\texttt{seg}_{k+1}$ happened to intersect all of them).
+
+To control this, **we first sort the segments by their left endpoints**. This means that each new segment can only either merge itself into a previous segment (without merging previously disjoint segments) or form a new disjoint segment.
```
Comment (jessechoe10) [inline solutions/gold/usaco-1018.mdx:140]:
```suggestion
To control this, **sort the segments by their left endpoints**. This means that each new segment can only either merge itself into a previous segment (without merging previously disjoint segments) or form a new disjoint segment.
```

**Review by jessechoe10 — CHANGES_REQUESTED** [review]

**Review by ryanchou-dev — APPROVED** [review]

**Pull request merged**